### DISCUSSION TITLE
Calling the constructor of a view from a DataTemplate

### DISCUSSION THREAD

**Discussion by 9vult** [AvaloniaUI/Avalonia#14584] [Q&A] [answered]
Hi, I've got a tab control being populated by a view. `Tabs` is an `ObservableCollection<TabViewModel>` that's updated in the viewmodel.

```xaml
<TabControl ItemsSource="{Binding Tabs}">
  <TabControl.ContentTemplate>
    <DataTemplate>
      <views:TabView  />
    </DataTemplate>
  </TabControl.ContentTemplate>
</TabControl>
```

This works great for displaying the content of each viewmodel in its respective tab. However, in the code-behind for TabView, I have some handler registrations:

```csharp
private async Task DoCopySelectedEventAsync(...) { }

public TabView()
{
  InitializeComponent();
  this.WhenActivated((CompositeDisposable disposables) =>
  {
    if (ViewModel != null)
    {
      ViewModel.CopySelectedEvents.RegisterHandler(DoCopySelectedEventAsync);
      [...]
    }
}
```

The issue I'm facing is the constructor for the TabView view is only called when it's first created, not for each instance. So the handlers only work in the first tab, and not in any of the other tabs.

With all that said, is there a way to get the TabView() constructor to be called for each entry in the DataTemplate so the handlers are set up for all of them? Thanks!

For reference: [The tabcontrol](https://github.com/9vult/Ameko/blob/fda6552c06c122e83f069e2661ec09e194380aef/Ameko/Views/MainView.axaml#L126), [The TabView code-behind](https://github.com/9vult/Ameko/blob/fda6552c06c122e83f069e2661ec09e194380aef/Ameko/Views/TabItemView.axaml.cs)

**Comment by maxkatz6** [comment, marked as answer]
TabView is likely getting reused. Which means, you probably should handle ViewModel changes.

Something like:
```csharp
  this.WhenActivated((CompositeDisposable disposables) =>
  {
     this.GetObservable(ViewModelProperty)
          .WhereNotNull()
          // Previous handler probably should be unregistered first
          .Subscribe(vm => vm.CopySelectedEvents.RegisterHandler(DoCopySelectedEventAsync))
          .DisposeWith(disposables);
  }
```

Haven't tested this code, but general idea is like that.

**Reply by 9vult** [reply]
Thanks, this does work!

Is there a method for unregistering? I know they're supposed to be unregistered on Dispose() but that's not applicable for tab switches since it's still loaded, just not visible.

One workaround for that would be to check if that vm has already been registered, and skip if that's the case, but if there's a better method I'd love to know!

**Reply by maxkatz6** [reply]
> that's not applicable for tab switches since it's still loaded, just not visible.

With TabControl specifically, only one tab content is loaded at a time. 

> One workaround for that would be to check if that vm has already been registered, and skip if that's the case, but if there's a better method I'd love to know!

I haven't worked with ReactiveUI in a while, so don't remember if it has a nice method for that, unfortunately.

**Comment by stevemonaco** [comment]
`TabControl.Content` is reused across the same type, so only one is made. Check https://github.com/AvaloniaUI/Avalonia/issues/14529 for more discussion. Consider moving to a combination of `TabStrip` instead for the tab selection and another control for the content that creates views for each item (eg. `ItemsControl` or `ContentControl` + your own caching).
